### PR TITLE
Fix slice_scatter

### PR DESCRIPTION
**Pull request by er3x3**
Current logic for slice_scatter doesn't handle the case when param start or end is out of range for [0, dim_value), while this case is valid. Most of the case (for transformer models) it will use MAX_INT for end. Current test data doesn't have such case so it doesn't cover such case. Using (torch.zeros(8, 8), torch.ones(2, 8), 0, 6, 100, 1) can fail current implementation.

This PR is to fix this, and also adjust the logic to make it simpler. The Transpose logic in current implementation is not necessary.

The new implementation passes all current test data as well as above one. I don't know how to add new test data to current UT code, someone please help me on this.

Fixes https://github.com/microsoft/onnxscript/issues/945

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -6247,40 +6247,21 @@ def aten_slice_scatter(
...
+        op.Unsqueeze(step, zero),
     )
-    # Generate index but not finalized, need to do transpose later
-    # e.g. [[0,1,2],[0,1,2],[0,1,2]...,[0,1,2]], total count = 2x3x7
-    index_base = op.Range(start, end, step)  # e.g. [0,1,2]
-    index_expand = op.Expand(index_base, shape_expand)
-    indices = op.Transpose(index_expand, perm=perm)
-
+    index_base = op.Unsqueeze(index_base, op.Range(1, op.Size(self_shape) - dim, 1))
+    shape_expand = op.ScatterElements(self_shape, op.Unsqueeze(op.Constant(value_int=dim), zero), one, axis=0)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6263]: Could you add a few lines to explain the logic for future readers? Thanks!

**Review by justinchuby — COMMENTED** [review]
Thanks for the clean implementation!

**Comment by justinchuby** [comment]
To add new tests, you may follow this: https://github.com/microsoft/onnxscript/blob/4d6dc45edce3c00c53acd8060b41c41501cde7d8/onnxscript/tests/function_libs/torch_lib/extra_opinfo.py#L754-L767

**Comment by justinchuby** [comment]
Does the PR fix this issue? https://github.com/microsoft/onnxscript/issues/945

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -6247,40 +6247,21 @@ def aten_slice_scatter(
...
-    dim_tensor = op.Reshape(op.Constant(value_int=dim), neg_1)
-    shape_before_dim = op.Slice(src_shape, op.Constant(value_ints=[0]), dim_tensor)
-    shape_after_dim = op.Slice(src_shape, op.Add(dim_tensor, 1), last_dim)
-    shape_expand = op.Concat(
-        shape_before_dim, shape_after_dim, op.Constant(value_ints=[1]), axis=0
+    zero = op.Constant(value_ints=[0])
+    one = op.Constant(value_ints=[1])
+    self_shape = op.Shape(self)
+    dim_shape = op.Gather(self_shape, dim, axis=0)
+    index_base = op.Range(0, dim_shape, 1)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6254]: Would be great to have a comment explaining how the case described in the PR is handled here

**Comment by fatcat-z** [comment]
> Does the PR fix this issue? #945

Looks like it does.

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py [hunk 3]
```diff
@@ -751,6 +751,131 @@ def sample_inputs_unfold(op_info, device, dtype, requires_grad, **kwargs):
...
+    )
+
+    L = 20
+    cases = (
+        (
+            (L, L, L),
+            (
+                L,
+                L,
+                L,
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:768]:
```suggestion
                L
```

nit: Remove the last comma to let black format the tuple into a compact line

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:768]: Below as well. Thanks!

**Review by justinchuby — APPROVED** [review]
Lgtm with comments

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py [hunk 4]
```diff
@@ -751,6 +751,35 @@ def sample_inputs_unfold(op_info, device, dtype, requires_grad, **kwargs):
...
+        ((L, L, L), (L // 4, L, L), (0, L // 2, L, 2)),
+        ((L, L, L), (L, L, L), (1, 0, L, 1)),
+        ((L, L, L), (L, L // 2, L), (1, L // 2, L, 1)),
+        ((L, L, L), (L, L // 4, L), (1, L // 2, L, 2)),
+        ((L, L, L), (L, L, L), (2, 0, L, 1)),
+        ((L, L, L), (L, L, L // 2), (2, L // 2, L, 1)),
+        ((L, L, L), (L, L, L // 4), (2, L // 2, L, 2)),
+        ((L, L, L), (L, L // 2, L), (1, L // 2, L * 2, 1)),  # end > L
+        ((L, L, L), (L, L, L), (-2, 0, L, 1)),  # negative dim
+        ((L, L, L), (L, L, L // 4), (-1, L // 2, L * 2, 2)),  # end > L and negative dim
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:774]: Nice comments, Thanks! Would you be willing to add these cases to pytorch as well?

Comment (er3x3) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:774]: > Nice comments, Thanks! Would you be willing to add these cases to pytorch as well?

I would rather not do this as I am not quite familiar with the process. You guys can do this. :-)

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:774]: The process is simple. It is just a matter of adding these three lines in a PR.

**Pull request merged**